torch/csrc/jit/passes/create_autodiff_subgraphs.cpp

```diff
@@ -1,257 +1,293 @@
-    const auto& vertex = dep_graph.at(ord);
-    if (!vertex) continue;
-    if (!shouldConsiderForMerge(vertex.value())) continue;
+  // Create a group node that contains only `n`
+  // `n` is destroyed.
+  Node* createSingletonGroup(Node* n) {
+    auto group = block_->owningGraph()->createDifferentiableSubgraph();
+    group->insertBefore(n);
+    mergeNodeIntoGroup(group, n);
+    return group;
+  }
 
-    auto& nodes = vertex.value()->data;
-    if (nodes.size() < size_threshold) continue;
+  value_list sortReverseTopological(ArrayRef<Value*> inputs) {
+    value_list result;
+    for (auto i : inputs) {
+      if (i->node()->owningBlock() == block_) {
+        result.push_back(i);
+      }
+    }
+    // Sort in reverse topological order
+    std::sort(result.begin(), result.end(), [&](Value* a, Value* b) {
+      return a->node()->isAfter(b->node());
+    });
+    return result;
+  }
 
-    diff_graphs.push_back(mergeNodes(block, prim::DifferentiableGraph, nodes));
+  bool shouldConsiderForMerge(Node* node) {
+    // if we're already in the process of merging
+    if (node->kind() == prim::DifferentiableGraph) {
+      return true;
+    }
+    if (node->kind() == prim::Constant) {
+      return false;
+    }
+    return isDifferentiable(node);
   }
-}
 
-void CreateAutodiffSubgraphsPK(
-    Block * block,
-    size_t size_threshold,
-    std::vector<Node*>& diff_graphs) {
-  for (auto * node : block->nodes()) {
-    // Find subgraphs to run this on recursively.
-    if (isDifferentiable(node)) continue;
-    for (auto * sub_block : node->blocks()) {
-      CreateAutodiffSubgraphsPK(sub_block, size_threshold, diff_graphs);
+  std::pair<graph_node_list::iterator, bool> scanNode(Node* consumer) {
+    if (shouldConsiderForMerge(consumer)) {
+      if (consumer->kind() != prim::DifferentiableGraph) {
+        consumer = createSingletonGroup(consumer);
+      }
+      auto inputs = sortReverseTopological(consumer->inputs());
+      for (auto input : inputs) {
+        if (auto group = tryMerge(consumer, input->node())) {
+          // we successfully merged, so the new group's `inputs` may have
+          // changed. So rescan the new group for more merging opportunities.
+          return std::make_pair(group.value()->reverseIterator(), true);
+        }
+      }
     }
+
+    return std::make_pair(++consumer->reverseIterator(), false);
   }
 
-  auto dep_graph = make_dependency_graph(block);
-  find_differentiable_groups(dep_graph);
-  reorder_according_to_dag(block, dep_graph);
-  merge_differentiable_groups(block, dep_graph, size_threshold, diff_graphs);
-}
+  // Try to merge `producer` into `consumer`. If successful, this destroys
+  // `producer` and returns the `consumer` group.
+  c10::optional<Node*> tryMerge(Node* consumer, Node* producer) {
+    JIT_ASSERT(consumer->kind() == prim::DifferentiableGraph);
+    bool canMerge = shouldConsiderForMerge(producer) &&
+        producer->moveBeforeTopologicallyValid(consumer);
```

Comment:
How much of the graph is `moveBeforeTopologicallyValid` allowed to change? Can it only touch the nodes between producer and consumer (topologically) or can it rewrite the entire graph?

Reply:
It can reorder any nodes between the source and destination. And nope, if a move is not possible the graph isn't changed.